File: typescript/test/client.test.ts
// Integration tests against the real san-db-ox binary (mirrors
// python/tests/test_client.py / go/sandbox/sandbox_test.go). Every test
// skips (does not fail) when no binary is found.

import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { Writable } from "node:stream";
import type { TestContext } from "node:test";
import { describe, test } from "node:test";
import type { Client, SocketClient } from "../src/index.js";
import {
  CODE_BAD_REQUEST,
  CODE_READ_ONLY,
  CODE_SQLITE_ERROR,
  connect,
  PROTOCOL,
  ResponseError,
  SanDBoxTimeoutError,
  VERSION,
} from "../src/index.js";
import { CALL_TIMEOUT_MS, copyOfBinary, repoRoot, sanDbOxBin } from "./helpers.js";

function requireBinary(t: TestContext): string | null {
  const bin = sanDbOxBin();
  if (!bin) {
    t.skip("no san-db-ox binary: run `make fetch` or set SAN_DB_OX_BIN");
    return null;
  }
  return bin;
}

async function openClient(bin: string, args: readonly string[] = []): Promise<Client> {
  return connect(bin, ["--serve-stdio", ...args], { timeoutMs: CALL_TIMEOUT_MS });
}

describe("connect", () => {
  test("hello reports protocol 1 and product SanDBox", {
    timeout: CALL_TIMEOUT_MS,
  }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const c = await openClient(bin);
    try {
      assert.equal(c.hello.protocol, PROTOCOL);
      assert.equal(c.hello.product, "SanDBox");
      assert.equal(typeof c.hello.protocol, "number");
    } finally {
      await c.close();
    }
  });

  test("a nonexistent command rejects rather than hanging", {
    timeout: CALL_TIMEOUT_MS,
  }, async () => {
    await assert.rejects(() =>
      connect("no-such-san-db-ox-binary-xyz", ["--serve-stdio"]),
    );
  });

  test("VERSION matches package.json", () => {
    const pkg = JSON.parse(
      readFileSync(join(repoRoot(), "typescript", "package.json"), "utf-8"),
    ) as { version: string };
    assert.equal(VERSION, pkg.version);
  });
});

describe("values", () => {
  test("BLOB round-trips through params, including empty", {
    timeout: CALL_TIMEOUT_MS,
  }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const c = await openClient(bin);
    try {
      await c.exec("CREATE TABLE t(b BLOB)");
      await c.exec("INSERT INTO t VALUES (?)", [new Uint8Array([1, 2, 3])]);
      await c.exec("INSERT INTO t VALUES (?)", [new Uint8Array()]);
      const r = await c.query("SELECT b FROM t ORDER BY rowid");
      assert.deepEqual(r.rows[0]?.[0], Buffer.from([1, 2, 3]));
      assert.deepEqual(r.rows[1]?.[0], Buffer.alloc(0));
    } finally {
      await c.close();
    }
  });

  test("a large integer round-trips as bigint", {
    timeout: CALL_TIMEOUT_MS,
  }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const c = await openClient(bin);
    try {
      await c.exec("CREATE TABLE t(x INTEGER)");
      await c.exec("INSERT INTO t VALUES (?)", [9223372036854775807n]);
      const r = await c.query("SELECT x FROM t");
      assert.equal(r.rows[0]?.[0], 9223372036854775807n);
    } finally {
      await c.close();
    }
  });

  test("REAL keeps a decimal point, and +-Inf/NaN round-trip", {
    timeout: CALL_TIMEOUT_MS,
  }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const c = await openClient(bin);
    try {
      const r = await c.query("SELECT 88.0");
      assert.equal(r.rows[0]?.[0], 88);
      assert.equal(typeof r.rows[0]?.[0], "number");

      // SQLite has no IEEE 754 division-by-zero semantics (1.0/0.0 is SQL
      // NULL, not Infinity) -- +-Inf only appears by overflowing a REAL
      // multiplication, exactly as conformance/cases/
      // real-and-integer-representation.json does.
      const inf = await c.query("SELECT 1e308 * 10");
      assert.equal(inf.rows[0]?.[0], Number.POSITIVE_INFINITY);
      const ninf = await c.query("SELECT -1e308 * 10");
      assert.equal(ninf.rows[0]?.[0], Number.NEGATIVE_INFINITY);
      // Inf - Inf is NaN, which the server sends as the JSON literal null
      // -- indistinguishable from SQL NULL once decoded (protocol.md).
      const nan = await c.query("SELECT (1e308 * 10) - (1e308 * 10)");
      assert.equal(nan.rows[0]?.[0], null);
    } finally {
      await c.close();
    }
  });
});

describe("errors", () => {
  test("stable error codes leave the connection usable", {
    timeout: CALL_TIMEOUT_MS,
  }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const c = await openClient(bin);
    try {
      await assert.rejects(
        () => c.exec("SELECT * FROM no_such_table"),
        (e: unknown) => e instanceof ResponseError && e.code === CODE_SQLITE_ERROR,
      );
      await assert.rejects(
        () => c.query("SELECT 1", [true as unknown as never]),
        // A TS caller can't construct this; a plain-JS caller can, and it
        // must fail client-side rather than reach the wire.
        TypeError,
      );
      await assert.rejects(
        // A missing required field ("sql") is dropped entirely by
        // JSON.stringify (undefined-valued properties are omitted), so
        // this reaches the wire as {"op":"query"} and exercises the
        // server's own bad_request path.
        // @ts-expect-error sql is intentionally omitted to trigger bad_request
        () => c.query(undefined),
        (e: unknown) => e instanceof ResponseError && e.code === CODE_BAD_REQUEST,
      );
      // Connection must still work after all of the above.
      const r = await c.query("SELECT 1");
      assert.equal(r.rows[0]?.[0], 1n);
    } finally {
      await c.close();
    }
  });
});

describe("--read-only", () => {
  test("op-level writes are rejected with read_only", {
    timeout: CALL_TIMEOUT_MS,
  }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const c = await openClient(bin, ["--read-only"]);
    try {
      await assert.rejects(
        () => c.snapshot(),
        (e: unknown) => e instanceof ResponseError && e.code === CODE_READ_ONLY,
      );
    } finally {
      await c.close();
    }
  });

  test("SQL-level writes are rejected by SQLite itself, as sqlite_error", {
    timeout: CALL_TIMEOUT_MS,
  }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const c = await openClient(bin, ["--read-only"]);
    try {
      await assert.rejects(
        () => c.exec("CREATE TABLE t(x)"),
        (e: unknown) => e instanceof ResponseError && e.code === CODE_SQLITE_ERROR,
      );
      // Reads still work; the process is still alive.
      const r = await c.query("SELECT 1");
      assert.equal(r.rows[0]?.[0], 1n);
    } finally {
      await c.close();
    }
  });
});

describe("snapshot / load / inspect / tables / schema / dump", () => {
  test("snapshot then load round-trips data across connections", {
    timeout: CALL_TIMEOUT_MS,
  }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const c1 = await openClient(bin);
    let path: string;
    try {
      await c1.exec("CREATE TABLE t(x)");
      await c1.exec("INSERT INTO t VALUES (1)");
      path = (await c1.snapshot()).path;
    } finally {
      await c1.close();
    }
    const c2 = await openClient(bin);
    try {
      await c2.load(path);
      const r = await c2.query("SELECT x FROM t");
      assert.equal(r.rows[0]?.[0], 1n);
    } finally {
      await c2.close();
    }
  });

  test("inspect describes only this process's own data", {
    timeout: CALL_TIMEOUT_MS,
  }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const c = await openClient(bin);
    try {
      const info = await c.inspect();
      assert.equal(typeof info.hasData, "boolean");
      assert.equal(typeof info.readOnly, "boolean");
      assert.equal(typeof info.source, "string");
    } finally {
      await c.close();
    }
  });

  test("tables / schema / dump", { timeout: CALL_TIMEOUT_MS }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const c = await openClient(bin);
    try {
      await c.exec("CREATE TABLE t(x)");
      const tables = await c.tables();
      assert.ok(tables.tables.includes("t"));
      const schema = await c.schema("t");
      assert.equal(schema.schema.length, 1);
      const dump = await c.dump();
      assert.ok(dump.sql.includes("CREATE TABLE t"));
    } finally {
      await c.close();
    }
  });
});

describe("close", () => {
  test("reaps the child and is idempotent", { timeout: CALL_TIMEOUT_MS }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const c = await openClient(bin);
    await c.close();
    assert.equal(c.exitCode, 0);
    await c.close();
    assert.equal(c.exitCode, 0);
  });
});

describe("overwrite", () => {
  test("reaps the child and persists data into a fresh connection", {
    timeout: CALL_TIMEOUT_MS,
  }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const { path: copyPath, cleanup } = copyOfBinary(bin);
    try {
      const c = await openClient(copyPath);
      await c.exec("CREATE TABLE t(x)");
      await c.exec("INSERT INTO t VALUES (42)");
      await c.overwrite();
      assert.equal(c.exitCode, 0);

      const c2 = await openClient(copyPath);
      try {
        const r = await c2.query("SELECT x FROM t");
        assert.equal(r.rows[0]?.[0], 42n);
      } finally {
        await c2.close();
      }
    } finally {
      cleanup();
    }
  });
});

describe("timeouts", () => {
  test("a timed-out call tears the connection down in the background", {
    timeout: CALL_TIMEOUT_MS,
  }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const c = await openClient(bin);
    try {
      await assert.rejects(
        () =>
          c.query(
            "WITH RECURSIVE cnt(x) AS (SELECT 1 UNION ALL SELECT x+1 FROM cnt WHERE x < 100000000) SELECT count(*) FROM cnt",
            [],
            { timeoutMs: 200 },
          ),
        SanDBoxTimeoutError,
      );
    } finally {
      // The connection is unusable after a timeout, but close() must
      // still be safe and must wait for the background teardown, so the
      // child is reliably reaped (no zombie left for the test runner).
      await c.close();
    }
    assert.notEqual(c.exitCode, null);
  });
});

describe("stderr sink", () => {
  test("wiring a stderr sink doesn't break the connection", {
    timeout: CALL_TIMEOUT_MS,
  }, async (t) => {
    const bin = requireBinary(t);
    if (!bin) return;
    const sink = new Writable({
      write(_chunk, _enc, cb) {
        cb();
      },
    });
    const c = await connect(bin, ["--serve-stdio"], {
      stderr: sink,
      timeoutMs: CALL_TIMEOUT_MS,
    });
    try {
      const r = await c.query("SELECT 1");
      assert.equal(r.rows[0]?.[0], 1n);
    } finally {
      await c.close();
    }
  });
});

describe("SocketClient has no direct-only APIs", () => {
  test("overwrite/exitCode are compile errors on SocketClient", () => {
    // A pure type-level check -- SocketClient's actual runtime behavior
    // (connecting over TCP/UNIX) is exercised in socket.test.ts. Here we
    // only need `make typescript-typecheck` to fail if either member is
    // ever accidentally added to SocketClient.
    const sc = {} as SocketClient;
    // @ts-expect-error SocketClient has no overwrite -- direct-only.
    void sc.overwrite;
    // @ts-expect-error SocketClient has no exitCode -- direct-only.
    void sc.exitCode;
  });
});

File: java/src/test/java/io/github/amisonnet8/sandbox/usage/ClientTest.java
package io.github.amisonnet8.sandbox.usage;

import io.github.amisonnet8.sandbox.ConnectionClosedException;
import io.github.amisonnet8.sandbox.ConnectOptions;
import io.github.amisonnet8.sandbox.ReadTimeoutException;
import io.github.amisonnet8.sandbox.ResponseException;
import io.github.amisonnet8.sandbox.SanDbOx;
import io.github.amisonnet8.sandbox.SanDbOxClient;
import io.github.amisonnet8.sandbox.SanDbOxException;
import io.github.amisonnet8.sandbox.SnapshotOptions;
import io.github.amisonnet8.sandbox.SnapshotResult;
import io.github.amisonnet8.sandbox.StderrSink;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.nio.file.Path;
import java.time.Duration;
import java.util.List;
import java.util.OptionalInt;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

/**
 * Integration tests against a real san-db-ox binary, using only the
 * public API. Package {@code usage}, separate from the rest of this test
 * suite, so that only {@link Connection}/{@link SanDbOx}-reachable code is
 * even compilable here -- the compiler is what enforces "this is
 * everything a consumer of this library could actually write" (mirrors
 * {@code rust/tests/client.rs} being a separate crate).
 *
 * <p>{@code bad_request} and {@code unsupported_op} are not reachable
 * through this typed API at all: there is no way to construct a malformed
 * request with it. Those two codes are covered by {@code
 * conformance/cases/error-codes.json} instead -- a type-safety win, not a
 * coverage gap.
 */
class ClientTest {

    private static SanDbOxClient open(Path bin) throws SanDbOxException {
        return SanDbOx.connect(bin.toString(), List.of("--serve-stdio"), new ConnectOptions().timeout(UsageTestSupport.CALL_TIMEOUT));
    }

    private static SanDbOxClient openWithArgs(Path bin, List<String> args) throws SanDbOxException {
        return SanDbOx.connect(bin.toString(), args, new ConnectOptions().timeout(UsageTestSupport.CALL_TIMEOUT));
    }

    private static Path requireBin() {
        Path bin = UsageTestSupport.sanDbOxBin();
        Assumptions.assumeTrue(bin != null, "skipping: no san-db-ox binary found: run `make fetch` or set SAN_DB_OX_BIN");
        return bin;
    }

    @Test
    void hello_and_basic_query() throws SanDbOxException {
        Path bin = requireBin();
        try (SanDbOxClient c = open(bin)) {
            assertEquals(SanDbOx.PROTOCOL, c.hello().protocol());
            assertEquals("SanDBox", c.hello().product());
            var result = c.query("SELECT 1", List.of());
            assertEquals(List.of(List.of(1L)), result.rows());
        }
    }

    @Test
    void connect_rejects_nonexistent_command() {
        // The spawn failure (ENOENT) surfaces as the base exception type:
        // there is no protocol violation or wire error to wrap here, just
        // a bad command.
        SanDbOxException e = assertThrows(
                SanDbOxException.class,
                () -> SanDbOx.connect("this-command-does-not-exist-really", List.of("--serve-stdio")));
        assertEquals(SanDbOxException.class, e.getClass());
    }

    @Test
    void blob_roundtrip() throws SanDbOxException {
        Path bin = requireBin();
        try (SanDbOxClient c = open(bin)) {
            c.exec("CREATE TABLE t(b BLOB)", List.of());
            c.exec("INSERT INTO t VALUES (?)", List.of((Object) "hi".getBytes(java.nio.charset.StandardCharsets.UTF_8)));
            var result = c.query("SELECT b FROM t", List.of());
            assertArrayEquals("hi".getBytes(java.nio.charset.StandardCharsets.UTF_8), (byte[]) result.rows().get(0).get(0));
        }
    }

    private static void assertArrayEquals(byte[] expected, byte[] actual) {
        org.junit.jupiter.api.Assertions.assertArrayEquals(expected, actual);
    }

    @Test
    void empty_blob_roundtrip() throws SanDbOxException {
        Path bin = requireBin();
        try (SanDbOxClient c = open(bin)) {
            c.exec("CREATE TABLE t(b BLOB)", List.of());
            c.exec("INSERT INTO t VALUES (?)", List.of((Object) new byte[0]));
            var result = c.query("SELECT b FROM t", List.of());
            assertArrayEquals(new byte[0], (byte[]) result.rows().get(0).get(0));
        }
    }

    @Test
    void large_integer_roundtrip() throws SanDbOxException {
        Path bin = requireBin();
        try (SanDbOxClient c = open(bin)) {
            long big = Long.MAX_VALUE;
            c.exec("CREATE TABLE big(n INTEGER)", List.of());
            c.exec("INSERT INTO big VALUES (?)", List.of(big));
            var result = c.query("SELECT n, typeof(n) FROM big", List.of());
            assertEquals(List.of(List.of(big, "integer")), result.rows());
        }
    }

    @Test
    void real_representation() throws SanDbOxException {
        Path bin = requireBin();
        try (SanDbOxClient c = open(bin)) {
            assertEquals(List.of(List.of(88.0)), c.query("SELECT 88.0", List.of()).rows());
            assertEquals(List.of(List.of(Double.POSITIVE_INFINITY)), c.query("SELECT 1e308 * 10", List.of()).rows());
            assertEquals(List.of(List.of(Double.NEGATIVE_INFINITY)), c.query("SELECT -1e308 * 10", List.of()).rows());
            // Inf - Inf is SQL NULL upstream, indistinguishable from a REAL
            // NaN once decoded -- not a loss introduced by this driver.
            List<List<Object>> nullRow = c.query("SELECT (1e308 * 10) - (1e308 * 10)", List.of()).rows();
            assertNull(nullRow.get(0).get(0));
        }
    }

    @Test
    void error_codes_and_connection_survives() throws SanDbOxException {
        Path bin = requireBin();
        try (SanDbOxClient c = open(bin)) {
            ResponseException e1 = assertThrows(ResponseException.class, () -> c.query("SELECT * FROM nope", List.of()));
            assertTrue(e1.isCode(SanDbOxException.CODE_SQLITE_ERROR), "got " + e1.code());

            ResponseException e2 = assertThrows(ResponseException.class, () -> c.load("/nonexistent/path/really-not-there.db"));
            assertTrue(e2.isCode(SanDbOxException.CODE_IO_ERROR), "got " + e2.code());

            // The connection must still be usable after errors.
            assertEquals(List.of(List.of(1L)), c.query("SELECT 1", List.of()).rows());
        }
    }

    @Test
    void read_only_two_tier_rejection() throws SanDbOxException {
        Path bin = requireBin();
        try (SanDbOxClient c = openWithArgs(bin, List.of("--serve-stdio", "--read-only"))) {
            // SQL-level write: rejected by SQLite's own PRAGMA query_only.
            ResponseException e1 = assertThrows(ResponseException.class, () -> c.exec("CREATE TABLE t(x INTEGER)", List.of()));
            assertTrue(e1.isCode(SanDbOxException.CODE_SQLITE_ERROR), "got " + e1.code());

            // Op-level writes: rejected by the stdio server itself, before SQLite.
            ResponseException e2 = assertThrows(ResponseException.class, c::overwrite);
            assertTrue(e2.isCode(SanDbOxException.CODE_READ_ONLY), "got " + e2.code());

            ResponseException e3 = assertThrows(ResponseException.class, () -> c.load("/nonexistent/does-not-matter.db"));
            assertTrue(e3.isCode(SanDbOxException.CODE_READ_ONLY), "got " + e3.code());

            assertEquals(List.of(List.of(1L)), c.query("SELECT 1", List.of()).rows());
            assertTrue(c.inspect().readOnly());
        }
    }

    @Test
    void snapshot_and_load() throws Exception {
        Path bin = requireBin();
        // snapshot()/load() write/resolve relative to the server's cwd --
        // both connections share one isolated directory so `path`
        // (returned by snapshot, possibly relative) resolves the same way
        // for load.
        Path cwd = UsageTestSupport.isolatedDir();
        try {
            String path;
            try (SanDbOxClient c = SanDbOx.connect(
                    bin.toString(), List.of("--serve-stdio"), new ConnectOptions().timeout(UsageTestSupport.CALL_TIMEOUT).cwd(cwd.toString()))) {
                c.exec("CREATE TABLE t(x INTEGER)", List.of());
                c.exec("INSERT INTO t VALUES (42)", List.of());
                SnapshotResult snap = c.snapshot(new SnapshotOptions().filename("snap"));
                path = snap.path();
            }

            try (SanDbOxClient c2 = SanDbOx.connect(
                    bin.toString(), List.of("--serve-stdio"), new ConnectOptions().timeout(UsageTestSupport.CALL_TIMEOUT).cwd(cwd.toString()))) {
                c2.load(path);
                assertEquals(List.of(List.of(42L)), c2.query("SELECT x FROM t", List.of()).rows());
            }
        } finally {
            UsageTestSupport.deleteRecursively(cwd);
        }
    }

    @Test
    void inspect_reflects_own_process_only() throws SanDbOxException {
        Path bin = requireBin();
        try (SanDbOxClient c = open(bin)) {
            var info = c.inspect();
            assertFalse(info.hasData());
            assertNull(info.dataLength());
            assertNull(info.version());
            c.exec("CREATE TABLE t(x INTEGER)", List.of());
            // exec'd SQL state doesn't change inspect's view of embedded data.
            var infoAfter = c.inspect();
            assertFalse(infoAfter.hasData());
        }
    }

    @Test
    void tables_schema_dump() throws SanDbOxException {
        Path bin = requireBin();
        try (SanDbOxClient c = open(bin)) {
            c.exec("CREATE TABLE users(id INTEGER PRIMARY KEY, name TEXT)", List.of());
            c.exec("CREATE TABLE logs(id INTEGER PRIMARY KEY, msg TEXT)", List.of());
            c.exec("INSERT INTO users(id, name) VALUES (1, 'alice')", List.of());

            assertEquals(List.of("logs", "users"), c.tables().tables());

            var schema = c.schema(null);
            assertEquals(
                    List.of(
                            "CREATE TABLE users(id INTEGER PRIMARY KEY, name TEXT)",
                            "CREATE TABLE logs(id INTEGER PRIMARY KEY, msg TEXT)"),
                    schema.schema());
            assertEquals(
                    List.of("CREATE TABLE users(id INTEGER PRIMARY KEY, name TEXT)"), c.schema("users").schema());

            var dump = c.dump("users");
            assertTrue(dump.sql().contains("INSERT INTO \"users\" VALUES(1,'alice')"));
            assertFalse(dump.sql().contains("logs"));
        }
    }

    @Test
    void close_reaps_process_and_is_idempotent() throws SanDbOxException {
        Path bin = requireBin();
        SanDbOxClient c = open(bin);
        c.close();
        assertEquals(OptionalInt.of(0), c.exitCode());
        c.close(); // must not throw or hang
    }

    @Test
    @SuppressWarnings("try") // the resource is deliberately unused in the body -- see the comment below
    void close_via_try_with_resources_reaps_child_without_explicit_close() throws SanDbOxException {
        Path bin = requireBin();
        try (SanDbOxClient ignored = open(bin)) {
            // closed automatically at the end of this block
        }
        // If the child leaked, there is nothing to assert on directly from
        // here (this test doesn't hold a handle to it past the block) --
        // the meaningful assertion is that the JVM running this test
        // completes cleanly without needing to force-kill anything, the
        // same practical signal the Rust driver's equivalent test relies on.
    }

    @Test
    void overwrite_reaps_and_persists_data() throws Exception {
        Path bin = requireBin();
        Path dir = UsageTestSupport.isolatedDir();
        try {
            Path copy = UsageTestSupport.copyOfBinary(bin, dir);

            try (SanDbOxClient c = open(copy)) {
                c.exec("CREATE TABLE t(x INTEGER)", List.of());
                c.exec("INSERT INTO t VALUES (7)", List.of());
                c.overwrite();
                assertEquals(OptionalInt.of(0), c.exitCode());
            }

            try (SanDbOxClient c2 = open(copy)) {
                assertTrue(c2.inspect().hasData());
                assertEquals(List.of(List.of(7L)), c2.query("SELECT x FROM t", List.of()).rows());
            }
        } finally {
            UsageTestSupport.deleteRecursively(dir);
        }
    }

    @Test
    void call_timeout_tears_down() throws SanDbOxException {
        Path bin = requireBin();
        // A plain "SELECT 1" round-trips too fast for a tiny timeout to
        // reliably fire -- use a query that takes a few seconds so the
        // timeout path is exercised deterministically rather than racing
        // real round-trip latency.
        String slowQuery = "WITH RECURSIVE r(x) AS (SELECT 1 UNION ALL SELECT x+1 FROM r WHERE x < 10000000) "
                + "SELECT count(*) FROM r";
        // Not try-with-resources: close() is called explicitly below as
        // the thing under test (that it's safe to call and that it
        // actually reaps the child after a timeout), so an additional
        // implicit close at the end of a resource block would be
        // redundant rather than a mistake -- same reason
        // close_reaps_process_and_is_idempotent above doesn't use it either.
        SanDbOxClient c = open(bin);
        c.setTimeout(Duration.ofMillis(200));
        assertThrows(ReadTimeoutException.class, () -> c.query(slowQuery, List.of()));

        // The connection is unusable now.
        assertThrows(ConnectionClosedException.class, () -> c.query("SELECT 1", List.of()));

        // close() must still be safe, and must have actually reaped the child.
        c.close();
        assertTrue(c.exitCode().isPresent());
    }

    @Test
    void stderr_sink_does_not_break_the_connection() throws SanDbOxException {
        Path bin = requireBin();
        ByteArrayOutputStream collected = new ByteArrayOutputStream();
        try (SanDbOxClient c = SanDbOx.connect(
                bin.toString(),
                List.of("--serve-stdio"),
                new ConnectOptions().timeout(UsageTestSupport.CALL_TIMEOUT).stderr(StderrSink.writer(collected)))) {
            c.query("SELECT 1", List.of());
            // Not asserting on content (san-db-ox may or may not log
            // anything for a clean run) -- just that wiring a sink doesn't
            // break the connection. A broken wiring would have made
            // connect() itself fail.
        }
    }

    @Test
    void nan_and_inf_params_are_rejected_before_anything_is_sent() throws SanDbOxException {
        Path bin = requireBin();
        try (SanDbOxClient c = open(bin)) {
            SanDbOxException e1 = assertThrows(SanDbOxException.class, () -> c.query("SELECT ?", List.of(Double.NaN)));
            assertTrue(e1.getMessage().contains("finite"), "got " + e1.getMessage());
            SanDbOxException e2 = assertThrows(SanDbOxException.class, () -> c.query("SELECT ?", List.of(Double.POSITIVE_INFINITY)));
            assertTrue(e2.getMessage().contains("finite"), "got " + e2.getMessage());
            // The connection is still usable: rejection happens
            // client-side, before a request line is ever written.
            assertEquals(List.of(List.of(1L)), c.query("SELECT 1", List.of()).rows());
        }
    }

    @Test
    void boolean_param_is_rejected() throws SanDbOxException {
        Path bin = requireBin();
        try (SanDbOxClient c = open(bin)) {
            SanDbOxException e = assertThrows(SanDbOxException.class, () -> c.query("SELECT ?", List.of(true)));
            assertTrue(e.getMessage().toLowerCase(java.util.Locale.ROOT).contains("boolean"), "got " + e.getMessage());
        }
    }
}
